### PR TITLE
Generate Entailment Probabilities

### PR DESCRIPTION
**Pull request by dleve123**
Addresses #72

### CODE REVIEW

**Review by anton164 — COMMENTED** [review]
Caught a minor bug, but otherwise looks good!

File: scripts/generate_all_entailment_labels.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+from datasets import load_dataset
+from sumtool.xsum_dataset import XsumDataset
+from typing import List, Tuple
+
+
+def load_entailment_model_and_tokenizer():
+    tokenizer = AutoTokenizer.from_pretrained("madlag/bert-large-uncased-mnli")
+    model = AutoModelForSequenceClassification.from_pretrained(
+        "madlag/bert-large-uncased-mnli"
+    ).to(device)
```
[the remark this answers is not in the document]
Comment (jxmorris12) [reply scripts/generate_all_entailment_labels.py:15]: wouldn't flake8 have caught this?

[the remark this answers is not in the document]
Comment (jxmorris12) [reply scripts/generate_all_entailment_labels.py:15]: oh @anton164 it's defined in the if block below so it's actually in scope here. Super confusing though haha

File: scripts/generate_all_entailment_labels.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        for batch_idx, batch in enumerate(tqdm(data_loader)):
+            batch_tokenized = tokenizer(
+                batch["document"],
+                batch["summary"],
+                padding=True,
+                return_tensors="pt",
+                truncation="only_first",
+            ).to(device)
+
+            batch_entailment_probs = (
```
Comment (jxmorris12) [inline scripts/generate_all_entailment_labels.py:84]: Just a small thing, if you wrap this in a `with torch.no_grad()` it'll be a good amount faster, and you won't have to call `detach()`.

File: scripts/generate_all_entailment_labels.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                .cpu()
+                .detach()
+                .numpy()
+                .tolist()
+            )
+
+            batch_entailment_metric = [
+                dict(zip(["entails_prob", "neutral_prob", "contradicts_prob"], probs))
+                for probs in batch_entailment_probs
+            ]
```
Comment (jxmorris12) [inline scripts/generate_all_entailment_labels.py:96]: another nit, this seems like a perfect place for [collections.namedtuple](https://docs.python.org/3/library/collections.html#collections.namedtuple)

**Review by jxmorris12 — APPROVED** [review]
LGTM! One more question, how did you pre-compute those hashes that are in the code? Seems a little unusual to do it this way

File: scripts/generate_all_entailment_labels.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (anton164) [reply scripts/generate_all_entailment_labels.py:15]: oh weird, well I would still suggest refactoring this to a method argument so it doesn't depend on the global scope

**Comment by anton164** [comment]
@jxmorris12 those hashes are hard-coded ids from the current storage implementation. It's not a great pattern so I've suggested a refactor which removes them here: https://github.com/factula/sumtool/pull/76 (it uses human-readable/user defined ids instead)

File: scripts/generate_all_entailment_labels.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dleve123) [reply scripts/generate_all_entailment_labels.py:15]: good catch! Will fix.

File: scripts/generate_all_entailment_labels.py (hunk 2 above)
Comment (dleve123) [reply scripts/generate_all_entailment_labels.py:84]: ah very cool - thanks.

File: scripts/generate_all_entailment_labels.py (hunk 3 above)
Comment (dleve123) [reply scripts/generate_all_entailment_labels.py:96]: I think I'm going to punt on this for now, for the sake of time!

**Comment by dleve123** [comment]
@jxmorris12 @anton164 changes addressed! If you one of you can give a quick 👍 , I'll merge

**Comment by jxmorris12** [comment]
Sorry @dleve123 I just realized one more issue! You need to call `model.eval` when loading in the model to deactivate dropout. This is important for reproducibility and also probably makes the results a bit better.

 (this bug applies to all of our evaluation code as well, hope we caught this elsewhere...)

**Comment by dleve123** [comment]
@jxmorris12 Hmm, seems like the docs suggest that [`from_pretrained` loads the model in eval mode](https://huggingface.co/docs/transformers/main_classes/model#transformers.PreTrainedModel.from_pretrained), which seems consistent with my poking around:

```python
(Pdb) list .
 52         device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
 53  
 54         tokenizer, model = load_entailment_model_and_tokenizer(device)
 55         breakpoint()
 56  
 57  ->     xsum_test_by_id = XsumDataset(load_dataset("xsum")["test"]).data_by_id
 58  
 59         generated_summary_lookup = {
 60             "BERTS2S": "2221cf6c3bad26c64a12417282dabf3c",
 61             "TranS2S": "adaab4ed6990eeae0ea8e98c78a722fe",
 62             "PtGen": "5641d358ff7a45cc65592938f1020675",
(Pdb) model.__dict__['training']
False
```

**Review by anton164 — APPROVED** [review]

**Comment by jxmorris12** [comment]
> @jxmorris12 Hmm, seems like the docs suggest that [`from_pretrained` loads the model in eval mode](https://huggingface.co/docs/transformers/main_classes/model#transformers.PreTrainedModel.from_pretrained), which seems consistent with my poking around:
> 
> ```python
> (Pdb) list .
>  52         device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
>  53  
>  54         tokenizer, model = load_entailment_model_and_tokenizer(device)
>  55         breakpoint()
>  56  
>  57  ->     xsum_test_by_id = XsumDataset(load_dataset("xsum")["test"]).data_by_id
>  58  
>  59         generated_summary_lookup = {
>  60             "BERTS2S": "2221cf6c3bad26c64a12417282dabf3c",
>  61             "TranS2S": "adaab4ed6990eeae0ea8e98c78a722fe",
>  62             "PtGen": "5641d358ff7a45cc65592938f1020675",
> (Pdb) model.__dict__['training']
> False
> ```

ok sounds good to me!
